javascript/src/URDFLoader.js

```diff
@@ -473,7 +473,7 @@ class URDFLoader {
                             .split(/\s/g)
                             .map(v => parseFloat(v));
 
-                    material.color.setRGB(rgba[0], rgba[1], rgba[2]);
+                    material.color = new THREE.Color(rgba[0], rgba[1], rgba[2]).convertSRGBToLinear();
```

Comment:
There's no need to create a new color here - `material.color.setRGB(rgba[0], rgba[1], rgba[2]).convertSRGBToLinear()` is good!

Reply:
Interesting - I am still kind of baffled by the ergonomics of Three's API sometimes - the copying and the in-place editing are strange. Will fix

Reply:
Where's the confusion? The solution involving creating a new Color instance requires modifying the new instance in place, as well.

Most of three.js' APIs either take a target to set the new values on to or modify the chained object in place. It's a necessity in the JS world to avoid unnecessary memory allocation that impacts the garbage collector.

Reply:
firstly, agreed that my solution creating a new Color is not ideal in terms of memory alloc. I wasn't clear on how to do it otherwise.

my main confusion with Three is often that it does exactly as you describe, modifying objects in place, which is the exact opposite of well-written component-model UIs like React apps (the application that surrounds our Three app, in my case). In React everything is immutable and must be replaced. In Three, everything is mutable and is expected to be edited. It's hard to context-switch, and sometimes Three's function names aren't clear (in my opinion) on whether they edit in-place or return a new object. For instance, I assumed `convertSRGBToLinear` does *not* edit in-place because it returns a `Color` - that leads me to believe that the return value is the desired outcome, not that it edits the object.

it's entirely possible that, as you say, this is the better way to do things in Three and it's just that I mostly operate in a different domain :smile:

Reply:
99% of the time new objects are not implicitly created especially with the low level math classes. `convertSRGBToLinear` returns the original instance of the color so it can be chained with other functions (similar to `set`) but perhaps the docs could be more clear. React does operate on a different paradigm but React is not necessarily designed to run at the rates that three.js typically does. Three.js WebGL animation typically runs at 60 frames per second so if new objects are created every frame memory allocation can add up fast and increase pressure on the GC which can lead to frame delays while the garbage collector works to release the memory which isn't great for interactive apps or games (especially VR). Three.js' API is designed to discourage constant creation of new objects and it's generally encouraged to create global reusable objects instead of new ones in functions when temp instances are needed.

Some of these functions used to return new objects but almost all of that functionality has been deprecated or removed. It might not be immediately intuitive especially if coming from other languages but it's a reality of Javascript and designed this way for good reasons. Of course it's up to you if you want to deal with that memory is a better way.

Reply:
the idea of passing references around all the time to save on memory (or in this case GC) is strikingly similar to the stuff we do in systems programming to be able to run in low-memory environments 😁
strange how some computing concepts manage to cross borders you wouldn't expect.